cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/DistributedSparkProgramRunner.java

```diff
@@ -0,0 +1,77 @@
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.internal.app.runtime.distributed;
+
+import co.cask.cdap.api.spark.Spark;
+import co.cask.cdap.api.spark.SparkSpecification;
+import co.cask.cdap.app.ApplicationSpecification;
+import co.cask.cdap.app.program.Program;
+import co.cask.cdap.app.runtime.ProgramController;
+import co.cask.cdap.app.runtime.ProgramOptions;
+import co.cask.cdap.app.runtime.ProgramRunner;
+import co.cask.cdap.common.app.RunIds;
+import co.cask.cdap.common.conf.CConfiguration;
+import co.cask.cdap.internal.app.runtime.ProgramOptionConstants;
+import co.cask.cdap.proto.ProgramType;
+import com.google.common.base.Preconditions;
+import com.google.inject.Inject;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.twill.api.RunId;
+import org.apache.twill.api.TwillController;
+import org.apache.twill.api.TwillRunner;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.util.Map;
+
+/**
+ * A {@link ProgramRunner} for launching {@link Spark} program in distributed mode. It starts
+ * a YARN application to act as the Spark client/driver. A secondary YARN application will be launched
+ * by Spark framework for running executor nodes.
+ */
+public class DistributedSparkProgramRunner extends AbstractDistributedProgramRunner {
+
+  private static final Logger LOG = LoggerFactory.getLogger(DistributedSparkProgramRunner.class);
+
+  @Inject
+  public DistributedSparkProgramRunner(TwillRunner twillRunner, Configuration hConf, CConfiguration cConf) {
+    super(twillRunner, hConf, cConf);
+  }
+
+  @Override
+  protected ProgramController launch(Program program, ProgramOptions options,
+                                     Map<String, File> localizeFiles, ApplicationLauncher launcher) {
+    // Extract and verify parameters
+    ApplicationSpecification appSpec = program.getApplicationSpecification();
+    Preconditions.checkNotNull(appSpec, "Missing application specification.");
+
+    ProgramType processorType = program.getType();
+    Preconditions.checkNotNull(processorType, "Missing processor type.");
+    Preconditions.checkArgument(processorType == ProgramType.SPARK, "Only SPARK process type is supported.");
+
+    SparkSpecification spec = appSpec.getSpark().get(program.getName());
+    Preconditions.checkNotNull(spec, "Missing SparkSpecification for %s", program.getName());
+
+    LOG.info("Launching Spark program: " + program.getName() + ":" + spec.getName());
```

Comment:
I think they are the same. Wasn't paying too much attention to the log since it's pretty much a copy from other distributed program runner. Will make a round of refactoring across all of them.